netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java

```diff
@@ -365,10 +437,22 @@ public void channelRegistered(ChannelHandlerContext ctx) throws Exception {
        * lifetime and we only want to configure it once.
        */
       if (handlers != null) {
-        ctx.pipeline().addFirst(handlers);
+        for (ChannelHandler handler : handlers) {
+          ctx.pipeline().addBefore(ctx.name(), null, handler);
+        }
+        ChannelHandler handler0 = handlers[0];
+        ChannelHandlerContext handler0Ctx = ctx.pipeline().context(handlers[0]);
         handlers = null;
+        if (handler0Ctx != null) { // The handler may have removed itself immediately
```

Comment:
> Does this comment apply to the case where handler0Ctx == null, where we don't need to call channelRegistered() on it?

Yes. `handler0Ctx == null` means the handler is not in the pipeline. Since it isn't in the pipeline, it shouldn't receive further events. I will note that it is possible that we may need to call `channelRegistered()` on a `handler1Ctx` instead, and that isn't handled in the code. The only time that `handler0Ctx == null` today is `plaintext()`, which completes immediately and does nothing.

`handler0Ctx.fireChannelRegistered()` (the context) does _not_ call `channelRegistered()` on `HttpProxyHandler` in your example. It calls the _next_ in the pipeline, since the context is intended for handlers to call to fire events, and it wouldn't want the event to notify itself.  Note that I also can't use the _previous_ context to call the handler, because there may not be a previous. This has been annoying multiple times.

So instead of using the context, most of the time we'll call the handler directly because it is a `ChannelInboundHandler` and so has the `channelRegistered()` method. That will trigger the event to go through the rest of the pipeline (including this handler, and since `handlers == null`, it'll continue the chain via `super.channelRegistered(ctx)`, which just calls `ctx.fireChannelRegistered()`). In the case that the handler doesn't have `channelRegistered()`, then using the context is fine because `handler0` doesn't care about the event anyway.

To be clear, I'm not happy with the complexity and knowledge required here, but the alternatives seemed to be worse (e.g., buggier and higher coupling). On the other hand, this is based on just Netty knowledge and the only bug I encountered while testing is when `handler0Ctx == null`.